javascriptlib/src/mill/javascriptlib/PublishModule.scala

```diff
@@ -0,0 +1,539 @@
+    val modDepsResources = moduleDeps.map { x => PathRef(x.millSourcePath / "resources") }
+    Seq(PathRef(millSourcePath / "resources")) ++ modDepsResources
+  }
+
+  /**
+   * Generate sources relative to publishDir / "typescript"
+   */
+  private def pubAllSources: T[IndexedSeq[String]] = Task {
+    val project = Task.workspace.toString
+    val fileExt: Path => Boolean = _.ext == "ts"
+    (for {
+      source <-
+        os.walk(sources().path) ++ pubModDepsSources().toIndexedSeq.flatMap(pr =>
+          os.walk(pr.path)
+        ).filter(fileExt)
+    } yield source.toString
+      .replaceFirst(millSourcePath.toString, "typescript")
+      .replaceFirst(project, "typescript")) ++ pubModDepsGenSources().map(pr =>
+      "typescript/generatedSources/" + pr.path.last
+    )
+
+  }
+
+  override def generatedSourcesPathsBuilder: T[Seq[(String, String)]] = Task {
+    Seq("@generated/*" -> "typescript/generatedSources")
+  }
+
+  override def upstreamPathsBuilder: Task[Seq[(String, String)]] = Task.Anon {
+
+    val upstreams = (for {
+      (res, mod) <- Task.traverse(moduleDeps)(_.resources)().zip(moduleDeps)
+    } yield {
+      Seq((
+        mod.millSourcePath.subRelativeTo(Task.workspace).toString + "/*",
+        s"typescript/${mod.millSourcePath.subRelativeTo(Task.workspace).toString}/src" + ":" + s"${declarationOut()}"
+      )) ++
+        res.map { rp =>
+          val resourceRoot = rp.path.last
+          val modName = mod.millSourcePath.subRelativeTo(Task.workspace).toString
+          // nb: resources are be moved in bundled stage
+          (
+            "@" + modName + s"/$resourceRoot/*",
+            resourceRoot match {
+              case s if s.contains(".dest") =>
+                rp.path.toString
+              case _ => s"typescript/$modName/$resourceRoot"
+            }
+          )
+        }
+
+    }).flatten
+
+    upstreams
+  }
+
+  override def typeRoots: T[ujson.Value] = Task.Anon {
+    ujson.Arr(
+      "node_modules/@types",
+      "declarations"
+    )
+  }
+
+  override def declarationDir: T[ujson.Value] = Task.Anon {
+    ujson.Str("declarations")
+  }
+
+  override def compilerOptionsPaths: Task[Map[String, String]] =
+    Task.Anon { Map.empty[String, String] }
+
+  override def compilerOptions: T[Map[String, ujson.Value]] = Task {
+    Map(
+      "declarationMap" -> ujson.Bool(true),
+      "esModuleInterop" -> ujson.Bool(true),
+      "baseUrl" -> ujson.Str("."),
+      "rootDir" -> ujson.Str("typescript"),
+      "declaration" -> ujson.Bool(true),
+      "outDir" -> ujson.Str(bundledOut()),
+      "plugins" -> ujson.Arr(
+        ujson.Obj("transform" -> "typescript-transform-paths"),
+        ujson.Obj(
+          "transform" -> "typescript-transform-paths",
+          "afterDeclarations" -> true
+        )
+      ),
+      "moduleResolution" -> ujson.Str("node"),
+      "module" -> ujson.Str("CommonJS"),
+      "target" -> ujson.Str("ES2020")
+    )
+  }
+
+  // patch typescript
```

Comment:
Do we really need to patch typescript to get rid of the `node_modules` prefix? Are there other ways to do it?

Reply:
- This is not to get rid of `node_modules` prefix, ts/js does not need the prefix to import node-modules, mill needs that. Node cannot resolve node_modules because we install them to `npmInstall.dest`, we have always needed the additional steps of pointing * to `npmInstall.dest` in our tsconfig, so it can resolve the modules, node expects your node_modules to be in the root of the project (or in a parent directory). we use the symlink to fix that.

- the node_modules/ prefix is also something not natural in js/ts development. I don't expect mill adopters for js/ts to prefix all their very many imports with node_module so they can use it.

- the ts-patch is to correctly resolve the ts-paths in our transpiled code, so they can run natively with node